src/irmin-pack/pack_key_intf.ml

```diff
@@ -17,19 +17,60 @@
 open! Import
 
 module type S = sig
-  type hash
-
-  include Irmin.Key.S with type t = hash and type hash := hash
+  include Irmin.Key.S
 
   val null : t
+
+  val unfindable_of_hash : hash -> t
+  (** [unfindable_of_hash h] is a key [k] such that [to_hash k = h], with an
+      unspecified internal representation. This function enables an efficient
+      implmentation of "portable" inodes, but is otherwise unused. Attempting to
+      dereference a key constructed in this way results in undefined behaviour. *)
 end
 
 module type Sigs = sig
-  type 'hash t = 'hash
+  type 'hash t
   (** The type of {i keys} referencing values stored in the [irmin-pack]
       backend. *)
 
-  module type S = S
+  (** The internal state of a key (read with {!inspect}).
+
+      Invariant: keys of the form {!Indexed} always reference values that have
+      entries in the index (as otherwise these keys could not be dereferenced). *)
```

Comment:
It's possible to build keys such that `find t k <> Some _` in a few ways:
- storage corruption: we don't check that the hash can be found in the index when decoding keys of `Indexed` type (and I don't think we should). For `Direct` keys, we _do_ do this (though only in order to match the previous eager integrity check behaviour).
- keys passed between different instances of the same store (i.e. from a writer instance to a reader one that is yet to `sync`). We support this explicitly for the direct case in `find_in_pack_file`, and implicitly via `Index.mem` reader semantics for `Indexed` keys.
- keys passed between different stores, which we obviously don't support.

I think these are still enough to consider this an "invariant", since an `Indexed` key passed to an out-of-date reader is still technically "in the index", the reader just needs to sync the index. Happy to change the wording though.

Is this the sort of detail you had in mind?

Reply:
I've added this explanation to the code.